### PR TITLE
Merge documentation

### PR DESCRIPTION
**Pull request by carmenromo**
Documentation has been included with the instructions to merge according to our requirements in IC.

### CODE REVIEW

File: doc/merging.rst
Comment (jacg) [inline doc/merging.rst]: This implies that *you* have approved the PR. Change to 'How to merge approved PRs'.

File: doc/merging.rst
Comment (jacg) [inline doc/merging.rst]: It is not a *requirement* that the remote be called `upstream` (you can call it whatever you like), but I'm reluctant to add this detail, as it would significantly lengthen a beautifully short line, and requiring it will probably help people from confusing themselves. So best leave it as it is.

`upstream` should be rendered in `code-face` (can't remember how it's done in RST), but the stars render it in *emphasis-face*. Same fix needed in lots more places throughout the document below.

File: doc/merging.rst
Comment (jacg) [inline doc/merging.rst]: ... make sure that it is ready to be merged ...

File: doc/merging.rst
Comment (jacg) [inline doc/merging.rst:37]: Probably should mention `--no-ff` and `-n`.

File: doc/merging.rst
Comment (jacg) [inline doc/merging.rst]: Is it described above?

File: doc/merging.rst
Comment (jacg) [inline doc/merging.rst]: Hmm, this is probably the most obscure git operation anyone on this project will ever do ... almost certainly the most obscure git operation that a budding NEXT merge-master has ever come across. As such, it could do with some explanation ... on the other hand, explaining it in text rather than interactively con peras seems hideously difficult, especially given the GitHub quirks where it sometimes only recognizes the author's id and sometimes only the full name, or email or whatever. Not sure what to do about this.

**Comment by jacg** [comment]
Oh, and @halmamol and @MiryamMV are, right now, the most qualified people on Earth to comment on what can/needs to be improved here!

**Comment by MiryamMV** [comment]
I'd include the explicit instructions for Magit (or CL). Perhaps they're not necessary for the experienced git users; I'm no expert and I'd appreciate the Magic/CL hints, specially for the `merge` and "author changing" bits.

**Comment by jacg** [comment]
> I'd include the explicit instructions for Magit (or CL).

On the one hand, I agree; on the other, this is a description of the expectations of merges in the IC repo, rather than a Git tutorial.

By adding many details about how Git is used, we dilute the message about IC merge requirements.

Perhaps we should do the following:

+ What is there so far, is not bad at all and provides a decent overview of the steps that need to be taken. Leave it as it is (apart from minor cosmetic changes).

+ Add a subsequent section giving detailed instructions for (some of) the steps, with references to these, from the corresponding parts of the overview.

**Comment by carmenromo** [comment]
> On the one hand, I agree; on the other, this is a description of the expectations of merges in the IC repo, rather than a Git tutorial.

Yes, I agree. I was not sure about all the explicit instructions, but it's true that it was for me an extremely useful tool. 

> Add a subsequent section giving detailed instructions for (some of) the steps, with references to these, from the corresponding parts of the overview.

Nice! I'll do it

**Comment by halmamol** [comment]
I think adding the magit commands would be useful, I'm still not too familiar with magit. The rest is fine! Thanks a lot @carmenromo !!

File: doc/merging.rst
Comment (carmenromo) [reply doc/merging.rst:37]: Do you think I should mention them here or in the commands instructions?

File: doc/merging.rst
Comment (carmenromo) [reply doc/merging.rst]: I don't understand what do you doubt here. To skip this step?

File: doc/merging.rst
Comment (jacg) [reply doc/merging.rst:37]: This is the overview. Details such as these should go in the detailed mechanics section.

File: doc/merging.rst
Comment (jacg) [reply doc/merging.rst]: Leave it as it is, other than adding a link to the detailed instructions how to achieve this, in the details section we agreed to add.

File: doc/merging.rst
Comment (mmkekic) [inline doc/merging.rst]: Safer to leave `reset` only in case the merger was erroneously developing in master?

File: doc/merging.rst
Comment (mmkekic) [reply doc/merging.rst]: Does this mean 'set the reviewer as the author of merge commit'?

File: doc/merging.rst
Comment (jacg) [reply doc/merging.rst]: If the merger was developing on `master` and resets it ... the work on it is lost! (Modulo commits that are in reflogs, yadda, yadda.)

Perhaps we should instruct the merger to create a `merge-pr` branch specifically for this purpose, then we don't have to make any assumptions about the local master.

**Comment by jacg** [comment]
For some reason I can't reply to the following in place:

> Does this mean 'set the reviewer as the author of merge commit'?

Yes.

File: doc/merging.rst
Comment (carmenromo) [reply doc/merging.rst]: Yes, to avoid that the merger person "receives the credits"

File: doc/merging.rst
Comment (carmenromo) [reply doc/merging.rst]: But this `merge-pr` branch would substitute the `master` branch or the branch created on top of the PR branch to merge?

File: doc/merging.rst
Comment (jacg) [reply doc/merging.rst]: Not so much about credits, as the fact that who reviewed and accepted it being more interesting and relevant information than who went through the mechanics of merging.

File: doc/merging.rst
Comment (jacg) [reply doc/merging.rst]: The process requires two branches, one tracking `upstream/master` and one tracking `author/feature`. It seems fairly safe for the merger to use a local `feature` to track `author/feature`. We're addressing the possibility raised by @mmkekic that the merger might have been doing something silly with local `master` so it would replace local `master`.

Perhaps it's not worth it. Perhaps we should just stick to `master`. I don't know, these are instructions for *people* and people are so complicated and do so many strange things! We can't possibly anticipate them all.

Comment (carmenromo) [reply doc/merging.rst]: I don't think it 's worth it. People should know that if they want to contribute, they should make a master independent branch :)

Comment (mmkekic) [reply doc/merging.rst]: Maybe simply saying 'Make sure your local `master` and `upstream/master` are pointing to the same commit' or some more precise/descriptive version of that?

Comment (carmenromo) [reply doc/merging.rst]: Yes, I like it!

**Comment by jacg** [comment]
Looks good.

I'd make it clear at the top that there are some detailed instructions at the bottom.

I'm not *entirely* sure about the correctness of all the commands, off the top of my head (playing them through in my head seemed to make sense), and I don't have the time to simulate the situation and check them all. If nobody else spots any mistakes, I suggest we go ahead. I suggest that the mergers try to verify the instructions during the next few merges they do. (Don't forget that you can see the rendered version of the document in the GitHub interface when looking at *this branch*, even before it is merged.)

The trouble is that, here, we're programming humans, so we can't write automated tests :-)

**Comment by halmamol** [comment]
> I'm not _entirely_ sure about the correctness of all the commands, off the top of my head (playing them through in my head seemed to make sense), and I don't have the time to simulate the situation and check them all. If nobody else spots any mistakes, I suggest we go ahead. I suggest that the mergers try to verify the instructions during the next few merges they do. (Don't forget that you can see the rendered version of the document in the GitHub interface when looking at _this branch_, even before it is merged.)

Would you prefer that @MiryamMV or I (who ever takes care of the next merge) uses this version as reference during the next merge before its approval? It would be the best way to cross check all the commands.

**Comment by jacg** [comment]
> Would you prefer that @MiryamMV or I (who ever takes care of the next merge) uses this version as reference during the next merge before its approval?

I was going to suggest this, but I was afraid that no new PR might come along for a while, and then we'd be left waiting. Adding a fix later, seemed the lesser evil. But if nobody objects to waiting until the new mergers have verified it on real PRs a few times, then I'm all for it.

**Comment by halmamol** [comment]
> I was going to suggest this, but I was afraid that no new PR might come along for a while, and then we'd be left waiting. Adding a fix later, seemed the lesser evil. But if nobody objects to waiting until the new mergers have verified it on real PRs a few times, then I'm all for it.

If you think that next PR is going to be in a while, then these instructions would be even more necessary. I wouldn't mind waiting. Up to you!

**Comment by carmenromo** [comment]
> I'm not entirely sure about the correctness of all the commands

I doubt in the `CLI` commands, the Magit ones I checked in there while I was writing them. But nice anyway if @halmamol and @MiryamMV make a second check!

> If you think that next PR is going to be in a while, then these instructions would be even more necessary. I wouldn't mind waiting. Up to you!

I wouldn't mind either!

**Comment by carmenromo** [comment]
> I'd make it clear at the top that there are some detailed instructions at the bottom.

Done!

**Comment by gonzaponte** [comment]
#785 is a tiny PR that could probably be reviewed, approved, and merged very quickly. It could be used for checking this documentation, but it needs a reviewer. I think my request for new reviewers can be declared a failure, so maybe @jacg or @mmkekic can take care of it?

**Comment by jacg** [comment]
> If you think that next PR is going to be in a while, then these instructions would be even more necessary.

... which is why I pointed out that these instructions are already readable in their beautifully-rendered glory in ... oh, let me be very specific this time ... [right here](https://github.com/carmenromo/IC/blob/merging-doc/doc/merging.rst).

That is to say, we don't need to merge this PR for its content to be available to the mergers. So we can take our sweet time in merging it, even if we *do* want to merge it *eventually*.

File: doc/merging.rst
Comment (MiryamMV) [inline doc/merging.rst]: Shouldn't it be:
**CLI:** ``git checkout -b <PR branchname> <author's remote>/<PR branchname>``  ?

Comment (carmenromo) [reply doc/merging.rst]: Yes! I'll fix it, thanks!

**Comment by halmamol** [comment]
Since we prefer Magit, we have revised the CLI with exception of the merge (we had a small issue and preferred to do it with Magit). Everything works fine!

**Comment by carmenromo** [comment]
> Since we prefer Magit, we have revised the CLI with exception of the merge (we had a small issue and preferred to do it with Magit). Everything works fine!

Nice! So, if @jacg agrees, this can be approved and merged!

**Comment by jacg** [comment]
Fine by me. (My `submit your review` button seems to have gone AWOL, so I can't give a formal approval.)

**Review by jacg — COMMENTED** [review]
Four people have pored over this a number of times, and the two new mergers have even performed some merges while verifying the instructions, so I think we can consider this to be as thoroughly tested as instructions-for-humans could reasonably be.

**Review by jacg — APPROVED** [review]
[Second attempt at *formal* approval]

Four people have pored over this a number of times, and the two new mergers have even performed some merges while verifying the instructions, so I think we can consider this to be as thoroughly tested as instructions-for-humans could reasonably be.

**Pull request merged**
